README.rst
pycnik
======


.. image:: https://secure.travis-ci.org/Mappy/pycnik.png
   :target: http://travis-ci.org/Mappy/pycnik
.. image:: https://pypip.in/v/pycnik/badge.png
   :target: https://crate.io/packages/pycnik
.. image:: https://pypip.in/d/pycnik/badge.png
   :target: https://crate.io/packages/pycnik
.. image:: https://d2weczhvl823v0.cloudfront.net/Mappy/pycnik/trend.png
   :alt: Bitdeli badge
   :target: https://bitdeli.com/free

A simple Translator from Python code (with coding conventions) to
Mapnik XML stylesheet.

- features:
    - allow using exotic number of zoom levels and tile sizes (computes scales denominators)
    - provides a mechanism for inheritance
    - automatically add `cache-feature` attribute when using more than 2 styles

- caveats:
    - still work to do to be compliant with https://github.com/mapnik/mapnik/wiki/XMLConfigReference


Install Pycnik
--------------

.. code-block:: bash

    $ git clone https://github.com/Mappy/pycnik.git
    $ cd pycnik
    $ python setup.py install

Or via pip::

    $ pip install pycnik

Dependencies:

    - python-mapnik (mapnik >= 2.x)
    - lxml

Testing
-------

Dependencies:

    $ pip install -r requirements/test.pip

To run the tests with nose:

    $ nosetests -v


Getting started
---------------

Pycnik uses dynamic variable declaration,
so you have to use the same keywords as the xml declaration syntax.

example.py:

.. code-block:: python

    from pycnik.model import *

    BACKGROUND_COLOR = 'rgb(255,255,220)'

    NATURAL_RASTER = {
        "type": "gdal",
        "file": "natural_earth.tif"
    }

    DATABASE_PARAM = {
        "dbname": "database",
        "estimate_extent": "true",
        "host": "0.0.0.0",
        "password": "******",
        "port": "5432",
        "type": "postgis",
        "user": "mapuser",
        "srid": "4326",
    }

    ################
    # MAP DEFINITION
    ################
    Map.background_color = BACKGROUND_COLOR
    Map.srs = "+init=epsg:4326"
    Map.minimum_version = "2.0"
    Map.font_directory = "fonts"
    Map.buffer_size = 128

    ########
    # LAYERS
    ########
    natural_earth = Layer("natural_earth")
    natural_earth.datasource = NATURAL_RASTER

    bnd = Layer("country boundaries")
    bnd.datasource = DATABASE_PARAM
    bnd.table = "schema.boundaries"

    ########
    # STYLES
    ########
    natural_earth.style()[:3] = {
        RASTER: {
            'scaling': 'bilinear'
        }
    }

    bnd.style("blue")[0:19] = {
        LINE: {
            'fill': 'rgb(255,0,0)',
            'stroke-width': '4'
        },
        'filter': "[countrycode]='ESP'"
    }

    # change the stroke width from level 10 to 15
    # the `filter` and `fill` attributes are preserved
    bnd.style("blue")[10:15] = {
        LINE: {'stroke-width': '12'}}


You can see more examples in the `test/resources <http://github.com/Mappy/pycnik/tree/master/test/resources>`_ directory.


Generate mapnik XML
-------------------

    $ pycnik example.py -o example.xml






CHANGELOG.rst
Changelog
=========

1.4 (2015-02-23)
-------------

- fix compute_scales for WGS84  

1.3.2 (2012-12-05)
------------------

- allowing any attributes for layer tag

1.3.1 (2012-10-25)
------------------

- Updated documentation


1.3 (2012-10-04)
----------------

- Removed "zoom" comment in tag scaledenominator (for mapnik 2.1)
- More tests


1.2 (2012-09-11)
----------------

- Added ``copy_style()`` helper


1.1 (2012-09-10)
----------------

- Support many symbolizer with the same type in rule
- More tests


1.0 (2012-08-07)
----------------

- Initial release



test/sample.py
from pycnik.model import *

Map.srs = "+proj=longlat +ellps=WGS84 +datum=WGS84 +no_defs"
Map.background_color = "steelblue"

world = Layer("world")
world.srs = "+proj=longlat +ellps=WGS84 +datum=WGS84 +no_defs"
world.datasource = {
    'type': 'shape',
    'file': 'ne_110m_admin_0_countries.shp'
}

world.style("My Style")[:] = {
    LINE: {
        'stroke': "rgb(50%,50%,50%)",
        'stroke-width': "0.1"
    }
}


test/__init__.py
# -*- coding: utf-8 -*-
from lxml import etree
from os.path import dirname, join
from StringIO import StringIO

from pycnik.pycnik import translate, import_style

RESOURCES = join(dirname(__file__), 'resources')


def resource(name):
    return join(RESOURCES, name)


def parse_resource(name):
    source = import_style(resource(name))
    output = StringIO(translate(source))
    return etree.parse(output)


test/test_map.py
# -*- coding: utf-8 -*-
import unittest
from nose.tools import *

from lxml import etree

from test import parse_resource, resource


MAP_XPATH = '/Map'


class TestMap(object):

    def setup(self):
        self.xml = parse_resource('sample.py')

    @unittest.skip("Mapnik DTD not up-to-date")
    def test_dtd(self):
        '''Should validate Mapnik DTD'''
        dtd = etree.DTD(open(resource('mapnik.dtd'), 'rb'))
        assert_true(dtd.validate(self.xml))

    def test_map_attribute(self):
        '''Should serialize map attributes'''
        assert_equal(len(self.xml.xpath(MAP_XPATH)), 1)
        maptag = self.xml.xpath(MAP_XPATH)[0]
        assert_equal(maptag.attrib['background-color'], 'steelblue')
        assert_equal(maptag.attrib['srs'], "+proj=longlat +ellps=WGS84 +datum=WGS84 +no_defs")
        assert_equal(maptag.attrib['minimum-version'], "2.0")


release.sh
#!/bin/sh
# Prompt for versions
# Clean-up release artifact
# Bump version and tag it
# Build a source distrbution and upload on PyPI
# Update version for new develpoment cycle
VERSION_FILE=pycnik/__init__.py

CURRENT=$(grep __version__ $VERSION_FILE | sed "s/__version__ = '\(.*\)'/\1/")
echo -n "Current version is $CURRENT, what version do you want to release ? "
read RELEASE


python setup.py clean
rm -rf *egg-info build dist
sed -i "s/$CURRENT/$RELEASE/" $VERSION_FILE
git commit $VERSION_FILE -m "Bump version $RELEASE"
git tag $RELEASE
python setup.py publish

echo -n "Version $RELEASE released, what version do you want for next development cycle ? "
read NEXT

sed -i "s/$RELEASE/$NEXT/" $VERSION_FILE
git commit $VERSION_FILE -m "Updated to version $NEXT for next development cycle"

echo "--------------------------------------------------------------"
echo "Released version $RELEASE and prepare $NEXT development cycle."


pycnik/model.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
"""
This module contains classes and symbolizers used in the python stylesheet

Symbolizer list available in Mapnik::

    Line (for lines & polygons)
    Polygon (for polygons)
    Point (for points)
    Text (for points, lines, and polygons)
    Shield (for points & lines)
    Line Pattern (for lines & polygons)
    Polygon Pattern (for polygons)
    Raster (for rasters)
    Markers (for points, lines, & polygons)
    Buildings

"""
from collections import OrderedDict
from copy import deepcopy

__all__ = (
    "Map", "Layer", "Style", "MetaWriter", "MetaCollector",
    "LINE", "POLYGON", "POINT", "TEXT",
    "SHIELD", "RASTER", "MARKERS", "BUILDING"
)

# list of available mapnik symbolizers
LINE = "line"
POLYGON = "polygon"
POINT = "point"
TEXT = "text"
SHIELD = "shield"
RASTER = "raster"
MARKERS = "markers"
BUILDING = "building"

SYMBOLIZERS = (
    LINE, POLYGON, POINT, TEXT,
    SHIELD, RASTER, MARKERS, BUILDING
)


def dict_merge(adict, bdict):
    """
    recursively merges dict's. not just simple a['key'] = b['key'], if
    both a and bhave a key who's value is a dict then dict_merge is called
    on both values and the result stored in the returned dictionary.

    """
    if not isinstance(bdict, dict):
        return bdict
    result = deepcopy(adict)
    for key, value in bdict.items():
        if key in result and isinstance(result[key], dict):
            result[key] = dict_merge(result[key], value)
        else:
            result[key] = deepcopy(value)
    return result


class Map(object):
    """
    used for carry Map Element attributes
    """
    # default values
    TILE_SIZE = 256
    LEVEL_NUMBER = 20
    ZOOM_FACTOR = 2
    srs = '+init=epsg:4326'

    def __init__(self, **kwargs):
        for elem, value in kwargs.items():
            setattr(self, elem, value)


class MetaWriter(object):
    """
    Metawriter tags defined at the top of the xml and used
    as symbolizer's attributes
    """
    def __init__(self, **kwargs):
        for elem, value in kwargs.items():
            setattr(self, elem, value)


class MetaCollector(object):
    """
    MetaCollector (replacement for MetaWriter made in Mappy)
    """
    def __init__(self, **kwargs):
        for elem, value in kwargs.items():
            setattr(self, elem, value)


class Style(tuple):
    """
    Represents a style within a mapnik layer.
    A Layer can have several styles, defined by their name.
    A style is a tuple containing a dictionnary for each level
    with symbolizer and attributes. ex:
    (
        {'line': {'fill': '#ffffff', ...}, 'point': {''}, # zoom 0
         'filter': "[name] = 'road'" },
        {'line': {'fill': '#ffffff', ...} }, # zoom 1
        {'line': {'fill': '#ffffff', ...} }, # zoom 2
    )
    """
    def __init__(self, iterable):
        super(Style, self).__init__(iterable)

    def __setitem__(self, key, val):
        """
        `key` could be an index value or a slice object.
        In case of a slice object, spawn dict to every levels.
        key.stop to simule real index
        """
        if isinstance(key, slice):
            start = key.start  # cause key is read only
            stop = key.stop

            if not stop:
                stop = len(self) - 1
            if not start:
                start = 0
            for lev in xrange(start, stop + 1):
                self[lev].update(dict_merge(self[lev], val))

        elif isinstance(key, int):
            self[key].update(dict_merge(self[key], val))


class Layer(object):
    """
    Class representing a mapnik Layer

    LayerInstance.style("stylename") creates a new Layer "stylename"

    styles = {
        "style1": style_instance, ...
    }

    """
    # painter contains layer instance
    # in order of declaration
    painter = []

    def __init__(self, name=""):
        self.name = name
        self.nlevels = Map.LEVEL_NUMBER
        self.styles = OrderedDict()
        self.__class__.painter.append(self)

    def __repr__(self):
        return "Layer: %s - nstyle: %d" % (self.name, len(self.styles))

    def style(self, stylename=""):
        """
        if stylename empty default style is created
        """
        if not stylename.strip():
            stylename = "default"

        stylename = "%s_%s" % (self.name, stylename)

        if stylename in self.styles:
            return self.styles[stylename]

        self.styles[stylename] = Style([{} for _ in xrange(self.nlevels)])
        return self.styles[stylename]


pycnik/pycnik.py
#!/usr/bin/python
# -*-encoding=utf-8-*-

# Copyright (C) <2012>  <Ludovic Delauné>
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
"""
Simple Translator from Python code to Mapnik XML stylesheet.

"""
import imp
import inspect
import pprint
from os.path import exists, join, dirname, abspath, isabs
from itertools import groupby

from lxml.etree import Element, SubElement, CDATA, tostring

try:
    import mapnik
except ImportError:
    import mapnik2 as mapnik

from model import SYMBOLIZERS, Map, MetaWriter, MetaCollector, Style, Layer

# assume that a pixel on a screen is 0.28mm on each side for metric projs
PIXEL_SIZE = 0.00028

# assume that a pixel on a screen is 3.55714704265e-09 on each side for latlon
# projs
PIXEL_SIZE_DEG= 3.55714704265e-09

def compute_scales(tile_size, nlevels, zoom_factor, srs):
    """
    returns a dictionnary with scales determined for each level based on
    `number of levels` (default is 20), `tile_size` (def 256px),
    the `zoom_factor`
    scales = {
        0: {'min': 73795.09, 'max': None},
        1: {'min': 24598.36, 'max': 73795.09},
        2: {'min': 8199.45, 'max': 24598.36},
        ...
    }
    """

    lonlat_proj = '+init=epsg:4326'

    prj = mapnik.Projection(srs)
    earth_width = abs(prj.forward(mapnik.Coord(-180, 0)).x * 2)
    scales = {}

    # lonlat projs
    if srs == lonlat_proj:
        pixel_size = PIXEL_SIZE_DEG
    # metric projs
    else:
        pixel_size = PIXEL_SIZE

    scales[0] = {"min": earth_width / (tile_size * pixel_size), 'max': None}

    # for other zoom, based on the zoom_factor
    for lev in range(1, nlevels):
        if lev != nlevels - 1:
            minscale = scales[lev - 1]['min'] / zoom_factor
        else:
            minscale = None

        scales[lev] = {
            'min': minscale,
            'max': scales[lev - 1]['min']
        }

    return scales

def checktype(value, typ):
    """
    raise a TypeError when `value` is not an instance of `typ`
    """
    if isinstance(value, typ):
        return
    raise TypeError("%s is not a %s" % (value, str(typ)))


def import_style(stylesheet):
    """
    import python style as module
    """
    if not exists(stylesheet):
        raise IOError("file not found")
    try:
        config = imp.load_source(stylesheet[:-3], stylesheet)
    except ImportError:
        print("Unable to import stylesheet file %s" % stylesheet)

    return config


def replace_underscore(value):
    return value.replace('_', '-')


def write_style(root, stylename, style, scales):
    """
    construct a style xml tree

    :param root: etree parent element
    :param stylename: name of the style
    :param style: instance of model.Style that contains rules etc
    :param scales: dict containing scales denominator
    """
    tagstyle = SubElement(root, "Style", name=stylename)

    # adding idx to levels and sorting by dict values
    # in order to group rules
    ordered_rules = sorted(zip(range(30), style), key=lambda elem: elem[1])

    grouped_rules = []
    # grouping with dict
    for k, g in groupby(ordered_rules, lambda elem: elem[1]):
        grouped_rules.append(k)
        # group: [(0, {'symbolizer': 'point'}),
        #         (1, {'symbolizer': 'point'})]
        group = list(g)

        dico = group[0][1]

        if not dico:
            # empty rule
            continue

        # firt level index
        first_level = group[0][0]
        last_level = first_level

        if len(group) > 1:
            # get last index
            last_level = group[-1][0]

        # rule tag
        rule = SubElement(tagstyle, "Rule")

        # scale denominator
        if last_level != len(style) - 1:
            SubElement(rule, "MinScaleDenominator").text = str(int(scales[last_level]['min']))
            if first_level != 0:
                SubElement(rule, "MaxScaleDenominator").text = str(int(scales[first_level]['max']))
        else:
            if first_level != 0:
                SubElement(rule, "MaxScaleDenominator").text = str(int(scales[first_level]['max']))

        for attr, value in dico.items():
            if attr in SYMBOLIZERS:
                if isinstance(value, list):
                    for subvalue in value:
                        symb = SubElement(rule, "%sSymbolizer" % attr.title())
                        for symatt, symval in subvalue.items():
                            symb.attrib[symatt] = str(symval)
                else:
                    checktype(value, dict)

                    symb = SubElement(rule, "%sSymbolizer" % attr.title())

                    for symatt, symval in value.items():
                        if symb.tag in ("ShieldSymbolizer", "TextSymbolizer")\
                                and symatt == 'value':
                            symb.text = CDATA(str(symval))
                            continue
                        symb.attrib[symatt] = str(symval)

            if attr == 'filter':
                filt = SubElement(rule, "Filter")
                filt.text = CDATA(str(value))

            if attr == 'linepattern':
                symb = SubElement(rule, "LinePatternSymbolizer")
                for symatt, symval in value.items():
                    symb.attrib[symatt] = str(symval)


def translate(source, output_file=None):
    """
    TODO: check styles duplicatas

    if `output_file` is None, returns xml as string
    """
    print("number of levels: %d" % source.Map.LEVEL_NUMBER)
    print("zoom factor: %d" % source.Map.ZOOM_FACTOR)
    print("using tile_size: %dpx" % source.Map.TILE_SIZE)

    scales = compute_scales(
        source.Map.TILE_SIZE, source.Map.LEVEL_NUMBER,
        source.Map.ZOOM_FACTOR, source.Map.srs
    )

    root = Element("Map")

    # writing Map tag and attributes
    for elem, value in source.Map.__dict__.items():
        if elem.startswith('__'):
            # skipping private var
            continue
        if elem in ('LEVEL_NUMBER', 'TILE_SIZE', 'ZOOM_FACTOR'):
            continue
        root.attrib[replace_underscore(elem)] = str(value)

    metawriters = [metaw for metaw in source.__dict__.itervalues()
        if isinstance(metaw, source.MetaWriter)]

    metacollectors = [metaw for metaw in source.__dict__.itervalues()
        if isinstance(metaw, source.MetaCollector)]

    # metawriters
    for meta in metawriters:
        metatag = SubElement(root, "MetaWriter")
        for attr, value in meta.__dict__.items():
            if attr.startswith('__'):
                # skipping private var
                continue
            metatag.attrib[replace_underscore(attr)] = value

    # metacollectors
    for meta in metacollectors:
        metatag = SubElement(root, "MetaCollector")
        for attr, value in meta.__dict__.items():
            if attr.startswith('__'):
                # skipping private var
                continue
            metatag.attrib[replace_underscore(attr)] = value

    # retrieve all instances of Layer
    layers = [layer for layer in source.__dict__.itervalues()
        if isinstance(layer, source.Layer)]

    # remove pointers to the same id
    layers = list(set(layers))
    # reorder style list
    layers = sorted(layers, key=lambda lay: source.Layer.painter.index(lay))

    print("Layers info : ")
    pprint.pprint(layers)

    # writing all styles at the top of the xml
    for lay in layers:
        for stylname, style in lay.styles.items():
            # making styles tags
            write_style(root, stylname, style, scales)

    # writing layers at the bottom
    for lay in layers:

        laytag = SubElement(root, "Layer", name=lay.name)

        # cache feature
        if len(lay.styles) >= 2:
            laytag.attrib["cache-features"] = 'on'
        # srs definition
        if getattr(lay, "srs", None):
            laytag.attrib["srs"] = lay.srs

        # style definition in order
        for stylname, style in lay.styles.items():
            stytag = SubElement(laytag, "StyleName")
            stytag.text = stylname

        # datasource parameters
        if getattr(lay, "datasource", None):
            datasrc = SubElement(laytag, 'Datasource')
            for name, value in lay.datasource.items():
                if name == "table":
                    continue
                param = SubElement(datasrc, "Parameter", name=name)
                param.text = value
            if getattr(lay, "table", None):
                SubElement(datasrc, "Parameter",
                    name="table").text = CDATA(lay.table)

        # get all other attributes
        for attr, value in lay.__dict__.items():
            if attr.lower() in ("styles", "srs", "datasource", "table", "nlevels"):
                continue
            if inspect.isbuiltin(getattr(lay, attr)) or attr.startswith('_'):
                continue
            laytag.attrib[replace_underscore(attr)] = str(value)

    if not output_file:
        return tostring(root,
            pretty_print=True,
            xml_declaration=True,
            encoding='utf-8')

    with open(output_file, 'wb') as out:
        out.write(tostring(root,
            pretty_print=True,
            xml_declaration=True,
            encoding='utf-8'))


def copy_style(filename, features=[], exclude=[]):
    '''
    Copy style from another stylesheet into the current one.

    :param filename: pycnik stylesheet filename to copy (must ends with .py)
    :param features: a feature list to copy. If empty, all features are copied.
    :param exclude: a feature list to exclude from copy.
    '''
    stack = inspect.stack()[1]
    caller = inspect.getmodule(stack[0])

    if not filename.endswith('.py'):
        raise ValueError("Stylesheet %s is not a python file" % filename)

    if not isabs(filename):
        filename = abspath(join(dirname(caller.__file__), filename))

    if not exists(filename):
        raise IOError("File %s not found" % filename)

    try:
        stylesheet = imp.load_source(filename[:-3], filename)
    except ImportError:
        print("Unable to import stylesheet file %s" % filename)

    if isinstance(features, str):
        features = [features]

    if not features:
        features = []
        for attr in dir(stylesheet):
            if inspect.isbuiltin(getattr(stylesheet, attr)) or attr.startswith('_'):
                continue

            if isinstance(attr, (str, int, float, dict, tuple, list, Map,
                    MetaWriter, MetaCollector, Style, Layer)):
                features.append(attr)

    for feature in features:
        if feature not in exclude:
            setattr(caller, feature, getattr(stylesheet, feature))


travis_mapnik.sh
#!/bin/bash
# Dirty ugly hack to use mapnik system installation

if [ -d "/usr/share/pyshared/mapnik" ]; then
    echo "Copying /usr/share/pyshared/mapnik into virtualenv"
    cp -R /usr/share/pyshared/mapnik $VIRTUAL_ENV/lib/python2.7/site-packages/
fi

if [ -d "/usr/share/pyshared/mapnik2" ]; then
    echo "Copying /usr/share/pyshared/mapnik2 into virtualenv"
    cp -R /usr/share/pyshared/mapnik2 $VIRTUAL_ENV/lib/python2.7/site-packages/
fi

if [ -d "/usr/lib/pyshared/python2.7/mapnik" ]; then
    echo "/usr/lib/pyshared/python2.7/mapnik/* into virtualenv"
    if [ -d "$VIRTUAL_ENV/lib/python2.7/site-packages/mapnik" ]; then
        mkdir -p $VIRTUAL_ENV/lib/python2.7/site-packages/mapnik
    fi
    cp /usr/lib/pyshared/python2.7/mapnik/* $VIRTUAL_ENV/lib/python2.7/site-packages/mapnik
fi

if [ -d "/usr/lib/pyshared/python2.7/mapnik2" ]; then
    echo "/usr/lib/pyshared/python2.7/mapnik2/* into virtualenv"
    if [ -d "$VIRTUAL_ENV/lib/python2.7/site-packages/mapnik2" ]; then
        mkdir -p $VIRTUAL_ENV/lib/python2.7/site-packages/mapnik2
    fi
    cp /usr/lib/pyshared/python2.7/mapnik2/* $VIRTUAL_ENV/lib/python2.7/site-packages/mapnik2
fi


examples/natural.py
from pycnik.model import *

BACKGROUND_COLOR = 'rgb(255,255,220)'

NATURAL_RASTER = {
    "type": "gdal",
    "file": "natural_earth.tif"
}

DATABASE_PARAM = {
    "dbname": "database",
    "estimate_extent": "true",
    "host": "0.0.0.0",
    "password": "******",
    "port": "5432",
    "type": "postgis",
    "user": "mapuser",
    "srid": "4326",
}

################
# MAP DEFINITION
################
Map.background_color = BACKGROUND_COLOR
Map.srs = "+init=epsg:4326"
Map.minimum_version = "2.0"
Map.font_directory = "fonts"
Map.buffer_size = 128

########
# LAYERS
########
natural_earth = Layer("natural_earth")
natural_earth.datasource = NATURAL_RASTER

bnd = Layer("country boundaries")
bnd.datasource = DATABASE_PARAM
bnd.table = "schema.boundaries"

########
# STYLES
########
natural_earth.style()[:3] = {
    RASTER: {
        'scaling': 'bilinear'
    }
}

bnd.style("blue")[0:19] = {
    LINE: {
        'fill': 'rgb(255,0,0)',
        'stroke-width': '4'
    },
    'filter': "[countrycode]='ESP'"
}

bnd.style("blue")[10:15] = {
    LINE: {
        'stroke-width': '12'
    }
    # inheritance, the filter is conserved
}
